File: packages/api/src/handlers/programs.ts
import { DeleteCommand, GetCommand, PutCommand, QueryCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyResultV2 } from 'aws-lambda';
import { docClient, getTableName, key } from '../db';

const tableName = () => getTableName();

export async function listPrograms(tenantId: string): Promise<APIGatewayProxyResultV2> {
  const pk = key.tenant(tenantId).pk;
  const result = await docClient.send(
    new QueryCommand({
      TableName: tableName(),
      KeyConditionExpression: 'pk = :pk AND begins_with(sk, :sk)',
      ExpressionAttributeValues: { ':pk': pk, ':sk': key.programSkPrefix },
    })
  );
  const items = (result.Items || []).map((item) => ({
    programId: item.sk?.replace(key.programSkPrefix, ''),
    name: item.name,
    currency: item.currency,
    earnRules: item.earnRules,
    burnRules: item.burnRules,
    tierConfig: item.tierConfig,
    createdAt: item.createdAt,
    updatedAt: item.updatedAt,
  }));
  return { statusCode: 200, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ programs: items }) };
}

export async function getProgram(tenantId: string, programId: string): Promise<APIGatewayProxyResultV2> {
  const { pk, sk } = key.program(tenantId, programId);
  const result = await docClient.send(
    new GetCommand({ TableName: tableName(), Key: { pk, sk } })
  );
  if (!result.Item) return { statusCode: 404, body: JSON.stringify({ error: 'Program not found' }) };
  const item = result.Item;
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      programId: item.sk?.replace(key.programSkPrefix, ''),
      name: item.name,
      currency: item.currency,
      earnRules: item.earnRules,
      burnRules: item.burnRules,
      tierConfig: item.tierConfig,
      createdAt: item.createdAt,
      updatedAt: item.updatedAt,
    }),
  };
}

export async function createProgram(tenantId: string, body: { name?: string; currency?: string }): Promise<APIGatewayProxyResultV2> {
  const programId = `prog_${Date.now()}_${Math.random().toString(36).slice(2, 9)}`;
  const now = new Date().toISOString();
  const { pk, sk } = key.program(tenantId, programId);
  await docClient.send(
    new PutCommand({
      TableName: tableName(),
      Item: {
        pk,
        sk,
        name: body.name || 'My Program',
        currency: body.currency || 'INR',
        earnRules: {},
        burnRules: {},
        tierConfig: {},
        createdAt: now,
        updatedAt: now,
      },
    })
  );
  return {
    statusCode: 201,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ programId, name: body.name || 'My Program', currency: body.currency || 'INR' }),
  };
}

export async function updateProgram(
  tenantId: string,
  programId: string,
  body: { name?: string; currency?: string; earnRules?: object; burnRules?: object; tierConfig?: object }
): Promise<APIGatewayProxyResultV2> {
  const { pk, sk } = key.program(tenantId, programId);
  const getResult = await docClient.send(new GetCommand({ TableName: tableName(), Key: { pk, sk } }));
  if (!getResult.Item) return { statusCode: 404, body: JSON.stringify({ error: 'Program not found' }) };
  const now = new Date().toISOString();
  const updated = {
    ...getResult.Item,
    ...(body.name !== undefined && { name: body.name }),
    ...(body.currency !== undefined && { currency: body.currency }),
    ...(body.earnRules !== undefined && { earnRules: body.earnRules }),
    ...(body.burnRules !== undefined && { burnRules: body.burnRules }),
    ...(body.tierConfig !== undefined && { tierConfig: body.tierConfig }),
    updatedAt: now,
  };
  await docClient.send(new PutCommand({ TableName: tableName(), Item: updated }));
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ programId, updatedAt: now }),
  };
}

File: packages/backend/src/app/services/programs.py
"""Program CRUD service."""

import time
from app.db import get_table, key
from app.exceptions import NotFoundError


def _table():
    return get_table()


def list_programs(tenant_id: str) -> list[dict]:
    pk = key.program_pk(tenant_id)
    result = _table().query(
        KeyConditionExpression="pk = :pk AND begins_with(sk, :sk)",
        ExpressionAttributeValues={":pk": pk, ":sk": key.program_sk_prefix()},
    )
    items = result.get("Items", [])
    return [
        {
            "programId": it["sk"].replace(key.program_sk_prefix(), ""),
            "name": it.get("name", ""),
            "currency": it.get("currency", "INR"),
            "earnRules": it.get("earnRules"),
            "burnRules": it.get("burnRules"),
            "tierConfig": it.get("tierConfig"),
            "createdAt": it.get("createdAt"),
            "updatedAt": it.get("updatedAt"),
        }
        for it in items
    ]


def get_program(tenant_id: str, program_id: str) -> dict | None:
    pk, sk = key.program(tenant_id, program_id)["pk"], key.program(tenant_id, program_id)["sk"]
    result = _table().get_item(Key={"pk": pk, "sk": sk})
    item = result.get("Item")
    if not item:
        return None
    return {
        "programId": item["sk"].replace(key.program_sk_prefix(), ""),
        "name": item.get("name", ""),
        "currency": item.get("currency", "INR"),
        "earnRules": item.get("earnRules"),
        "burnRules": item.get("burnRules"),
        "tierConfig": item.get("tierConfig"),
        "createdAt": item.get("createdAt"),
        "updatedAt": item.get("updatedAt"),
    }


def create_program(tenant_id: str, name: str, currency: str) -> dict:
    program_id = f"prog_{int(time.time() * 1000)}_{__import__('random').randbytes(4).hex()}"
    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    pk, sk = key.program(tenant_id, program_id)["pk"], key.program(tenant_id, program_id)["sk"]
    _table().put_item(
        Item={
            "pk": pk,
            "sk": sk,
            "name": name,
            "currency": currency,
            "earnRules": {},
            "burnRules": {},
            "tierConfig": {},
            "createdAt": now,
            "updatedAt": now,
        },
    )
    return {"programId": program_id, "name": name, "currency": currency}


def update_program(
    tenant_id: str,
    program_id: str,
    *,
    name: str | None = None,
    currency: str | None = None,
    earn_rules: dict | None = None,
    burn_rules: dict | None = None,
    tier_config: dict | None = None,
) -> dict:
    existing = get_program(tenant_id, program_id)
    if not existing:
        raise NotFoundError("Program not found")
    now = time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())
    pk, sk = key.program(tenant_id, program_id)["pk"], key.program(tenant_id, program_id)["sk"]
    item = {
        "pk": pk,
        "sk": sk,
        "name": name if name is not None else existing["name"],
        "currency": currency if currency is not None else existing["currency"],
        "earnRules": earn_rules if earn_rules is not None else existing.get("earnRules", {}),
        "burnRules": burn_rules if burn_rules is not None else existing.get("burnRules", {}),
        "tierConfig": tier_config if tier_config is not None else existing.get("tierConfig", {}),
        "createdAt": existing.get("createdAt", now),
        "updatedAt": now,
    }
    _table().put_item(Item=item)
    return {"programId": program_id, "updatedAt": now}
